src/utils/log_config.py

```diff
@@ -0,0 +1,178 @@
+    :author: Adam Lozinsky
+
+    :example:
+        if __name__ == '__main__':
+            from utils import log_config
+            log = log_config.log(log_level='INFO')
+            log.info(f"Example info text {[1, 2, 3]}", example_key=[1, 2, 3])
+            try:
+                main()
+            except Exception as exec:
+                log.exception("Example crashed", exception=exec)
+
+    :notes:
+        Setting up structlog is very tricky, but after it is done it should just work
+        (or so they promise). Regardless, the comments herein should help explain how this
+        works should we ever need to update it.
+
+        See documentation.
+        https://docs.python.org/3/library/logging.handlers.html#timedrotatingfilehandler
+        https://www.structlog.org/en/stable/standard-library.html
+        https://www.structlog.org/en/stable/processors.html#chains
+"""
+
+import os
+import json
+import inspect
+from pathlib import Path
+import sys
+# We need these two handlers from logging to print to a file and stdout
+import logging
+from logging import StreamHandler
+from logging.handlers import TimedRotatingFileHandler
+import structlog
+import graypy
+# We need rich to make the console look pretty (Requires Python 3.7+)
+import rich
+from rich import pretty
+rich.pretty.install()
+
+
+def swap_logger_name(_, __, event_dict):
+    """
+    Swaps the kw 'logger_name' value with the 'module_name' and 'func_name' values then removes them
+    from the 'event_dict' for nicer representation. This is done to hack ConsoleRenderer to somewhat
+    match our past format. This is intended only to be used for console rendering and not
+    JSON rendering (prints nice but does not appear in file).
+    """
+
+    event_dict["logger"] = event_dict["module"] + " " + event_dict["func_name"]
+    del event_dict["module"]
+    del event_dict["func_name"]
+
+    return event_dict
+
+
+def log(log_level=None, console=None, logfile=None, aggregator=None):
+    """
+    :param log_level: Logging threshold [CRITICAL, ERROR, WARNING, INFO, DEBUG, NOTSET]
+    :type log_level: str
+    :param console: Enable (True) or Disable (False) console logging override
+    :type console: bool
+    :param logfile: Enable (True) or Disable (False) JSON file logging override
+    :type logfile: bool
+    :param aggregator: Enable (True) or Disable (False) aggregator log forwarding override
+    :type aggregator: bool
+
+    :notes:
+    There are three parts to logging; processors, renderers, and handlers.
+        processors - modify, add, or clean up the log message dict
+        renderers - make the log message a string, json, dict, etc. with fancy styling
+        handlers -  print the rendered data to stdout, file, stream, etc.
+    """
+
+    # Obtain the module name that imported this log_config
+    caller = Path(inspect.stack()[-1].filename)
+    module_name = caller.name.split('.')[0]
+
+    # Gather the borealis configuration information
```

Comment:
In the future this could be moved to a single borealis-wide options file that reads the config file

Reply:
Not sure what you mean here? Do you mean the whole try: with open(config).... part? Because, yes, that is duplicated all over the place.

Reply:
I have added in src.utils.general the load_config() and load_restrict() functions. I updated the code everywhere that we load configs to now use these functions. The caller part remains unchaged as it is only needed in log_config.py.